File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        i = 0

        while i < len(s):
            c = s[i]
            match = {')': '(', '}': '{', ']': '['}.get(c)

            if match and (not stack or stack.pop() != match) or not match:
                stack.append(c)
        
            i += 1

        return not stack
    

File: 20-valid-parentheses/valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const stack = [];
    let i = 0;

    while (i < s.length) {
        const c = s[i++];
        const match = { ')': '(', '}': '{', ']': '[' }[c];

        if (match && stack.pop() !== match || !match) {
            stack.push(c);
        }
    }

    return stack.length === 0;
};
